_posts/2020-06-30-sparklyr-1.3.0-released/sparklyr-1.3.0.Rmd

```diff
@@ -0,0 +1,192 @@
+
+sdf <- sdf_copy_to(
+  sc,
+  tibble::tibble(
+    a = c(1, NaN, 3, 4, NaN),
+    b = c(-2L, 0L, 1L, 3L, 2L),
+    c = c("a", "b", "c", "", "d")
+  )
+)
+
+# This example Avro schema is a JSON string that essentially says all columns
+# ("a", "b", "c") of `sdf` are nullable.
+avro_schema <- jsonlite::toJSON(list(
+  type = "record",
+  name = "topLevelRecord",
+  fields = list(
+    list(name = "a", type = list("double", "null")),
+    list(name = "b", type = list("int", "null")),
+    list(name = "c", type = list("string", "null"))
+  )
+), auto_unbox = TRUE)
+
+# persist the Spark data frame from above in Avro format
+spark_write_avro(sdf, "/tmp/data.avro", as.character(avro_schema))
+
+# and then read the same data frame back
+spark_read_avro(sc, "/tmp/data.avro")
+
+```
+
+```
+# Source: spark<data> [?? x 3]
+      a     b c
+  <dbl> <int> <chr>
+  1     1    -2 "a"
+  2   NaN     0 "b"
+  3     3     1 "c"
+  4     4     3 ""
+  5   NaN     2 "d"
+
+```
+
+## Custom Serialization
+
+In addition to commonly used data serialization formats such as CSV, JSON, Parquet, and Avro, starting from sparklyr 1.3, customized data frame serialization and deserialization procedures implemented in R can also be run on Spark workers via the newly implemented `spark_read` and `spark_write` methods. We can see both of them in action through a quick example below, where `saveRDS` is called from a user-defined writer function to save all rows within a Spark data frame into 2 RDS files on disk, and `readRDS` is called from a user-defined reader function to read the data from the RDS files back to Spark:
+
+```{r echo=TRUE, eval=FALSE}
+library(sparklyr)
+
+sc <- spark_connect(master = "local")
+sdf <- sdf_len(sc, 7)
+paths <- c("/tmp/file1.RDS", "/tmp/file2.RDS")
+
+spark_write(sdf, writer = function(df, path) saveRDS(df, path), paths = paths)
+spark_read(sc, paths, reader = function(path) readRDS(path), columns = c(id = "integer"))
+```
+
+```
+# Source: spark<?> [?? x 1]
+     id
+  <int>
+1     1
+2     2
+3     3
+4     4
+5     5
+6     6
+7     7
+```
+
+## Sparklyr.flint
+
+[Sparklyr.flint](https://github.com/r-spark/sparklyr.flint) is a sparklyr extension that aims to make functionalities from the [Flint](https://github.com/twosigma/flint) time series library easily accessible from R. It is currently under active development. One good news is while the original [Flint](https://github.com/twosigma/flint) library was designed to work with Spark 2.x, a slightly modified [fork](https://github.com/yl790/flint) of it will work well with Spark 3.0, and within the existing sparklyr extension framework, sparklyr.flint can automatically determine which version of the Flint library to load based on the version of Spark it's connected to. Another good news is, as previously mentioned, sparklyr.flint doesn't know too much about its own destiny yet. Maybe you can play an active part in shaping its future!
```

Comment:
I think it probably still makes more sense to briefly mention `sparklyr.flint` in this blog post, describing it as "under active development", so that the broader open source community (esp. those who work extensively with time series in R) will be aware of it early on and hopefully can give us some insight into how the R interface should look like and what functionalities they really want to get out of `sparklyr.flint`.

Also, I hope some of our audience will like the thinly veiled Back to the Future reference from above.